db/view/view.cc

```diff
@@ -2623,7 +2626,137 @@ future<> view_builder::migrate_to_v2(locator::token_metadata_ptr tmptr, db::syst
+        if (!_contexts[shard].semaphore) {
+            _contexts[shard].semaphore = &_db.local().get_reader_concurrency_semaphore();
+        }
+        return *_contexts[shard].semaphore;
+    }
+    virtual future<reader_permit> obtain_reader_permit(schema_ptr schema, const char* const description, db::timeout_clock::time_point timeout, tracing::trace_state_ptr trace_ptr) override {
+        return make_ready_future<reader_permit>(semaphore().make_tracking_only_permit(schema, description, timeout, std::move(trace_ptr)));
+    }
+};
+
+void view_builder::init_virtual_table() {
+    if (_view_build_status_on_group0) {
+        // We set the old system_distributed.view_build_status table to read virtually
+        // from system.view_build_status_v2 in order to make the transition transparent for
+        // readers of the table and maintain compatibility.
+
+        auto ms_v2 = mutation_source([this] (schema_ptr s,
+                reader_permit permit,
+                const dht::partition_range& pr,
+                const query::partition_slice& ps,
+                tracing::trace_state_ptr trace_state,
+                streamed_mutation::forwarding,
+                mutation_reader::forwarding fwd_mr) {
+
+            // The v1 distributed table and the v2 system local table have different shard mapping.
+            // The v2 table uses the null sharder, everything is on shard zero, while the v1 table
+            // has the normal sharder.
+            // So to simulate a read in v1 table in some shard, we need to read all the rows
+            // belonging to this shard from v2 table in shard zero.
+            // In order to read the table in a different shard we use the multishard reader. It is
+            // set to read from all the shards, but actually it only has rows to read in shard zero.
+            // Then we also need to filter only the keys belonging to this shard according to v1 sharder.
+
+            auto& table_v2 = _db.find_column_family(db::system_keyspace::view_build_status_v2());
+
+            mutation_reader ms_reader = make_multishard_combining_reader_v2(
+                    seastar::make_shared<view_build_status_policy>(_db.container(), table_v2.schema()->id()),
+                    table_v2.schema(),
+                    table_v2.get_effective_replication_map(),
+                    std::move(permit),
+                    pr, ps, trace_state, fwd_mr);
+
+            auto& sharder_v1 = s->get_sharder();
+            auto filter_fn = [&sharder_v1, shard_id = this_shard_id()] (const dht::decorated_key& dk) {
+                return sharder_v1.shard_for_reads(dk.token()) == shard_id;
+            };
+
+            return make_filtering_reader(std::move(ms_reader), std::move(filter_fn));
+        });
+
+        auto& table_v1 = _db.find_column_family(db::system_distributed_keyspace::NAME, db::system_distributed_keyspace::VIEW_BUILD_STATUS);
+        table_v1.set_virtual_reader(std::move(ms_v2));
```

Comment:
This is a pretty unusual virtual table, I think it is the first one which pretends to be a distributed table, all existing virtual tables use local replication strategy.
Since the underlying data source is a distributed one (local table replicated via raft), this should actually work just fine: replicas will receive requests that they "own" (based on how vnodes are distributed).

You have to watch out for repairs: both read-repair and regular repair. Both have code to ignore local replicated tables, but they might attemp to repair this table, which might have unexpected results. A read repair can happen if a read happens to step on a partition, which was replicated to one replica, but not yet to the other one, at the time of the read. I think the read repair will fail because virtual tables do not accept writes. I don't know if this is a problem or not in practice. A workaround would be to add a write handler too (see `system.config` virtual table), which just silently swallows writes. Maybe that can hide problems though, where some buggy code tries to write the v1 table.

Reply:
No, the writes will fail, try manually writing into any virtual table (including this new one).

Reply:
I tried writing to the table and it succeeds, except I can't read it back.
It's not initialized like other virtual tables. It's a normal distributed table with a virtual reader. I didn't set a virtual writer that throws an exception, like in other virtual tables, but I can add it.

Reply:
I changed it to set a virtual writer to ignore writes to the table